### PR TITLE
Project scheme

### PR DESCRIPTION
**Pull request by fortmarek**
### Short description 📝

When we were trying to get rid of `Workspace.swift`, there was one thing we could not replicate without it: having a single scheme that would run *all* tests that are defined in the current `Graph`. 

### Solution 📦

Taking inspiration from SPM which creates `xxx-Package` we create `xxx-Project` that contains all targets for building/testing from the current graph. 

### Implementation 👩‍💻👨‍💻

- [x] Create `xxx-Project` scheme
- [x] Add tests
- [x] Add changelog

### CODE REVIEW

**Review by kwridan — COMMENTED** [review]
This a nice idea, having a workspace wide scheme with all targets for testing, we've manually defined something similar in our `Workspace.swift`. Thanks for pushing this @fortmarek 👍 

A few high level thoughts on the approach _(feel free to reach out, happy to assist / discuss)_

### Workspace Mappers

* +1 for the introduction of Workspace mappers, it was a key missing mapper!
* It doesn't seem it's used in this PR though?
* I'd make a slight tweak to its signature:

```swift
func map(workspace: Workspace, projects: [Project]) throws -> (Workspace, [Project], [SideEffectDescription])
```

As those mappers are run ahead of having a graph. Once we have a graph, things get a lot harder to map _(internal inconsistencies can easily occur)_.
 
I can push something to assist with this shortly, see if that helps to make the workspace mapper work here instead of the graph mapper.

### Schemes on Graph

* `schemes` on `graph` could be a bit error prone in its current form
  * It doesn't get populated when loading the graph standalone outside of `TuistKit.ProjectGenerator`
  * It gets populated via one of the mappers only if a config option is enabled
  * Lastly, only used in one of the generation methods _(e.g. for consistency, it would be good for it to work when using `loadWorkspace` too)_

* From what I gathered looking through the code, there is another motivation / use-case for the `graph.schemes` and that is to use it within `BuildGraphInspector`
   * The idea to extract all schemes from a graph is still sound, what I'd suggest is perhaps rather than make the graph the storage container for them, it can contain a computed accessor to extract them from their source of truth (`Project`) - otherwise we'd have to keep them in sync :/

```swift
class Graph {
   // ...
   var schemes: [Scheme] {
        graph.projects.flatMap(\.schemes)
   }
}
```

**Comment by fortmarek** [comment]
Thanks for the thorough feedback @kwridan!

### Workspace Mappers 
I see you already have made a draft PR for `WorkspaceMapper`s - I have used those initially and forgot to delete it because they are, indeed, not used currently. So, I'll delete those and let's move the discussion about them to that PR 🙂 

### Schemes on Graph

> Lastly, only used in one of the generation methods (e.g. for consistency, it would be good for it to work when using loadWorkspace too)

Good catch, I'll do that!

As for the other points:
I can see that schemes can be error-prone in its current form, but they can not be completely computed either - because in order to make the new scheme work, I need to cross-reference projects which is only possible in `Workspace`. What I can do, though, is add another property called `additionalSchemes` where we could add schemes that can not be added to projects. The schemes property would then look something like this:

```
var schemes: [Scheme] {
  graph.project.flatMap(\.schemes) + additionalSchemes
}
```

That should fix most of your points, what do you think, @kwridan?

**Comment by kwridan** [comment]
> I need to cross-reference projects which is only possible in Workspace

ah good point, missed this bit. Thinking out loud, what if we add `Workspace` to `graph`? 

```swift
class Graph {
  let workspace: Workspace?
  
  // ...
  var schemes: [Scheme] {
      return projects.flatMap(\.schemes) + workspace?.schemes ?? []
  }
}
```

_(requires a small shuffle of where the `Workspace` is created when using `loadProjectWorkspace` - did some of that in #1767)_

That way we ensure they are always in sync. The reason for my hesitance to store `schemes` on graph is to avoid potential errors, where other components read or update those properties, and not all components using them consistently (e.g. the schemes set there may not be generated if not set on `Workspace` too etc...)

> That should fix most of your points, what do you think, @kwridan?

What you proposed should address project scheme consistency, but we'll still have potential inconsistencies for the workspace sadly. Possibly the mapper to introduce the Project scheme can be written as a workspace mapper using #1767 + the addition above may be able to solve both requirements while keeping consistency?

- Automatically generate a workspace level scheme to include all targets
- Make workspace schemes available within the graph (so it can be used by other components like `BuildGraphInspector`)

**Comment by fortmarek** [comment]
Hmm, adding `Workspace` to the `Graph` should work reliably!

Let's first merge [#1767](https://github.com/tuist/tuist/pull/1767/) and I'll then incorporate your suggested changes, @kwridan 🙂

**Comment by fortmarek** [comment]
Soo, I have hit a bit of a roadblock and I'd appreciate some ideas.

The scheme works great for running all test schemes that it can find.

It also works well for building all the available targets in the current `Graph` in the fixture it's being currently run for (`app_with_framework_and_tests`).

But when I tried to run it on our company project, I have stumbled upon an unfortunate problem.

Our simplified setup is as follows:
```
- MainApp
- Features
      - FeatureA
           - FeatureFrameworkA
           - ExampleAppA
      - FeatureB
           - FeatureFrameworkB
           - ExampleAppB
....
```

When the `MainApp-Project` scheme is generated it contains all the available targets - including `MainApp` and `ExampleAppA`, `B`, etc. When I try to build this scheme, I get a lot of errors that tell me:
```
Multiple commands produce "...Build/Products/Debug-iphonesimulator/ReactiveSwift.framework.dsym"
```

Now, all these app schemes are runnable targets and they all have `Embed precompiled frameworks` build phase. This build phase has specified `Output files` where for each framework there are these four following rows:
```
${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/ReactiveSwift.framework
${DWARF_DSYM_FOLDER_PATH}/ReactiveSwift.framework.dSYM
${BUILT_PRODUCTS_DIR}/236DDF56-E190-3A07-B509-BB01A127320C.bcsymbolmap
${BUILT_PRODUCTS_DIR}/F45E9641-612C-3752-9D00-D0F76BF614EC.bcsymbolmap
``` 

The first one is fine since it's specific for a given target. But unfortunately, `DWARF_DSYM_FOLDER_PATH` and `BUILT_PRODUCTS_DIR` are the same throughout the app whatever the target is. That means if `MainApp` and `ExampleAppA` both have eg. `ReactiveSwift`, they will have the same output files for `.dsym` and `.bcysmbolmap`. Therefore Xcode fails the build because it's able to analyze these conflicts.

There are two possible solutions that I thought of, none of which are perfect:
- We don't specify output files for `Embed precompiled frameworks` build phase. If I understand correctly, that'd unfortunately mean that it'd copy all the symbols on each build which would increase build times
- We modify `xxx-Project` scheme so it can not be used for *building* all targets - and it'd only include all testable targets. This'd still bring a lot of value to users since running all testable targets with one scheme is the reason why most users use `Workspace.swift` and this would alleviate that need. On the other hand *building* all targets has some value, too - eg. we would like to easily verify that all of our example apps are buildable and this new scheme would make this possible for us. 

As it stands now, I'll probably be forced to delete the functionality of building all targets, but maybe there is a better solution somewhere.

Would you, @kwridan or @pepibumur, have any ideas? Anything is appreciated 🙏

**Comment by kwridan** [comment]
Ouch looks like you've hit https://github.com/tuist/tuist/issues/919

> - We don't specify output files for Embed precompiled frameworks build phase. If I understand correctly, that'd unfortunately mean that it'd copy all the symbols on each build which would increase build times.

We'll need to re-check the current script and how it relies on input/output paths, but perhaps limiting what is specified rather than removing all output paths, e.g. keeping `${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/ReactiveSwift.framework` which is unique as you mentioned. 

When using Carthage, I believe their recommendation is to only specify the `.framework` and `carthage copy-frameworks` command will still copy the dsyms under the hood if present. This approach will be somewhat consistent with that - worth a try :) 

> - We modify xxx-Project scheme so it can not be used for building all targets - and it'd only include all testable targets. This'd still bring a lot of value to users since running all testable targets with one scheme is the reason why most users use Workspace.swift and this would alleviate that need. On the other hand building all targets has some value, too - eg. we would like to easily verify that all of our example apps are buildable and this new scheme would make this possible for us.

Indeed this would still add value, if I'm not mistaken we may still hit the same error in some scenarios. For example, if a test target has a host application set, I believe Xcode will build both the test target and the host application (or is that purely controlled via the scheme?). In case any of the host applications end up embedding the same precompiled framework we'd end up with the same issue :/ 

To my understanding `Workspace.swift` is not going anywhere and is still a very valid solution which allows you to create custom cross-project schemes. We use it to create several schemes, `AllUnitTests`, `AllUITests`, `FeatureUITests`,  etc... As much as automating the generation of those was appealing for us, we always found a need for some custom control over which targets to group in a scheme, by type, domain, team etc...

Hope this helps

**Comment by thedavidharris** [comment]
> When using Carthage, I believe their recommendation is to only specify the .framework and carthage copy-frameworks command will still copy the dsyms under the hood if present. This approach will be somewhat consistent with that - worth a try :)

Yeah, that's how the copy script handles it. CocoaPods had this same error at one point, fixed in these PRs: https://github.com/CocoaPods/CocoaPods/pull/9547/files + https://github.com/CocoaPods/CocoaPods/pull/9552.

DSYMs don't really need to get "embedded" like frameworks, so it's potentially better/possible to treat it fully separate.

> To my understanding Workspace.swift is not going anywhere and is still a very valid solution which allows you to create custom cross-project schemes. We use it to create several schemes, AllUnitTests, AllUITests, FeatureUITests, etc... As much as automating the generation of those was appealing for us, we always found a need for some custom control over which targets to group in a scheme, by type, domain, team etc...

This would be something good to maybe put as a future abstraction. One thing that we have been attempting is to include all things under `Libraries/**` in the Workspace.swift and have that be part of the AllTests scheme which is the most efficient way to build on CI. However, the best end goal I see here is the ability to have something like a `tuist test --changed-modules` which can generate a `ChangedTestTargets` scheme on the fly and have xcodebuild build and test all these in parallel. This PR does get most of the plumbing at a build level to get there.

**Comment by fortmarek** [comment]
Thanks @kwridan and @thedavidharris for the feedback!

> When using Carthage, I believe their recommendation is to only specify the .framework and carthage copy-frameworks command will still copy the dsyms under the hood if present. This approach will be somewhat consistent with that - worth a try :)

I suppose it does make sense to copy the Carthage approach - I have refactored the code and now the scheme is buildable even when there multiple products with `Embed precompiled frameworks` build phase 🎉 

> However, the best end goal I see here is the ability to have something like a tuist test --changed-modules which can generate a ChangedTestTargets scheme on the fly and have xcodebuild build and test all these in parallel

This is something we can definitely achieve! I am working on a `test` command and this is something I'd love to integrate. Would you mind filing an issue for this feature, so it's tracked somewhere?

**Comment by thedavidharris** [comment]
Tested this out on our main project, and it works quite well. Successfully picked up every scheme as defined and was able to build/test all of them much more efficiently.

**Review by kwridan — APPROVED** [review]
nice work @fortmarek 👍

File: Sources/TuistKit/ProjectGenerator/ProjectGenerator.swift [hunk 1]
```diff
@@ -232,7 +232,12 @@ class ProjectGenerator: ProjectGenerating {
             .workspace
             .merging(projects: updatedGraph.projects.map { $0.path })
 
-        return (updatedWorkspace, project, updatedGraph, modelMapperSideEffects + graphMapperSideEffects)
+        return (
```
Comment (pepibumur) [inline Sources/TuistKit/ProjectGenerator/ProjectGenerator.swift:235]: With 4 returned, I think it's time to group them together in a struct? That way we can access those elements using a more descriptive name (e.g. `returnedType.project`) as opposed to `returnedType.1`.

File: Sources/TuistKit/ProjectGenerator/ProjectGenerator.swift [hunk 2]
```diff
@@ -267,7 +272,11 @@ class ProjectGenerator: ProjectGenerating {
         let updatedWorkspace = workspace
             .merging(projects: updatedGraph.projects.map { $0.path })
 
-        return (updatedWorkspace, updatedGraph, modelMapperSideEffects + graphMapperSideEffects)
+        return (
```
Comment (pepibumur) [inline Sources/TuistKit/ProjectGenerator/ProjectGenerator.swift:275]: Similar comment here in regarding to the tuple, I'd return a `struct` instead.

**Review by pepibumur — APPROVED** [review]
Solid work @fortmarek 👏

**Pull request merged**